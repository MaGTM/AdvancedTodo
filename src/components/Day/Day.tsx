import React, {useMemo, useState} from 'react';
import {useNavigate, useParams} from "react-router-dom";
import {Day as DayModel} from '../../models/Day'
import {Backdrop, Box, Stack, Typography} from "@mui/material";
import CloseIcon from '@mui/icons-material/Close';
import ArrowBackIcon from '@mui/icons-material/ArrowBack';
import DayItem from "./DayItem/DayItem";
import DayModal from "./DayModal/DayModal";

const Day = () => {
    let {date} = useParams()
    let navigate = useNavigate()
    let [isModalOpen, setIsModalOpen] = useState<{ isOpen: boolean, hour: string }>({isOpen: false, hour: ''})

    let day = useMemo(() => {
        return new DayModel(date!)
    }, [])

    let closeHandler = (e?: React.MouseEvent) => {
        if ((e?.target as Element).getAttribute('data-backdrop') ||
            ((e?.target as Element).parentElement!.getAttribute('data-backdrop') && (e?.target as Element).nodeName === 'path')) {
            setIsModalOpen({isOpen: false, hour: ''})
        }

    }

    return (
        <>
            <Stack spacing={4}>
                <Box sx={{
                    border: '2px #fff solid',
                    borderRadius: '50%',
                    width: '50px',
                    height: '50px',
                    display: 'flex',
                    alignItems: 'center',
                    pl: '5px',
                    cursor: 'pointer'
                }}
                     onClick={() => navigate(-1)}
                >
                    <ArrowBackIcon sx={{color: '#fff'}} fontSize={'large'}/>
                </Box>
                <Typography variant={'h3'} component={'h1'} color={'#fff'}>
                    {day.getDay()} {day.getMonth()} {day.getYear()}
                </Typography>
                <Stack spacing={2}>
                    {day.getDayHours().map(item => {
                        return <DayItem key={item.hour} hour={item} setIsModalOpen={setIsModalOpen}/>
                    })}
                </Stack>
            </Stack>
            <Backdrop open={isModalOpen.isOpen} onClick={closeHandler} data-backdrop={true}>
                <DayModal day={day} hour={isModalOpen.hour} closeHandler={setIsModalOpen}/>
                <CloseIcon sx={{position: 'absolute', top: '15px', right: '15px', color: '#fff', cursor: 'pointer'}}
                           fontSize={'large'} data-backdrop={true}/>
            </Backdrop>
        </>
    );
};

export default Day;