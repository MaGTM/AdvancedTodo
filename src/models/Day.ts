import {monthToString} from "../components/utils/monthToString";
import {IDay, IDayHours} from "./IDay";

export class Day implements IDay{
    readonly day: number
    readonly month: number
    readonly year: number
    dayHours: IDayHours[] = []

    constructor(date: string) {
        this.day = Number(date.split('-')[0])
        this.month = Number(date.split('-')[1])
        this.year = Number(date.split('-')[2])
    }

    getHour(number: number) {
        if(number < 10) {
            return `0${number}:00`
        }
        if(number >= 10) {
            return `${number}:00`
        }
        return ''
    }

    getDay() {
        return this.day
    }

    getMonth() {
        return monthToString(this.month-1)
    }

    getYear() {
        return this.year
    }

    getDayHours() {
        if(this.dayHours.length > 0) return this.dayHours

        if(localStorage.getItem(`${this.day}-${this.month}-${this.year}`)) {
            this.dayHours = JSON.parse(localStorage.getItem(`${this.day}-${this.month}-${this.year}`)!)
            return this.dayHours
        }

        for(let i = 0; i <= 24; i++) {
            this.dayHours.push({hour: this.getHour(i), todos: []})
        }
        return this.dayHours
    }

    setHourTodo(hour: string, data: {todo: string, isUrgent: boolean}) {
        if(!data.todo) return
        this.dayHours[this.dayHours.findIndex(el => el.hour === hour)].todos.push(data)
        localStorage.setItem(String(`${this.day}-${this.month}-${this.year}`), JSON.stringify(this.dayHours))
    }
}