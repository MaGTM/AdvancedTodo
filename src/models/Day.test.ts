import {Day} from "./Day";


describe('Calendar Model', () => {
    test('Day', () => {
        const day = new Day(`22-11-2022`)
        expect(day.getDay()).toBe(22)
    })
    test('Hour', () => {
        const day = new Day(`22-11-2022`)
        expect(day.getHour(11)).toBe('11:00')
    })
    test('Hour', () => {
        const day = new Day(`22-11-2022`)
        expect(day.getHour(9)).toBe('09:00')
    })
    test('Year', () => {
        const day = new Day(`22-11-2022`)
        expect(day.getYear()).toBe(2022)
    })
    test('Month', () => {
        const day = new Day(`22-11-2022`)
        expect(day.getDayHours()).toMatchSnapshot()
    })
})
