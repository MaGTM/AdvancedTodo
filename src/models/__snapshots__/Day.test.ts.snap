// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar Model Month 1`] = `
Array [
  Object {
    "hour": "00:00",
    "todos": Array [],
  },
  Object {
    "hour": "01:00",
    "todos": Array [],
  },
  Object {
    "hour": "02:00",
    "todos": Array [],
  },
  Object {
    "hour": "03:00",
    "todos": Array [],
  },
  Object {
    "hour": "04:00",
    "todos": Array [],
  },
  Object {
    "hour": "05:00",
    "todos": Array [],
  },
  Object {
    "hour": "06:00",
    "todos": Array [],
  },
  Object {
    "hour": "07:00",
    "todos": Array [],
  },
  Object {
    "hour": "08:00",
    "todos": Array [],
  },
  Object {
    "hour": "09:00",
    "todos": Array [],
  },
  Object {
    "hour": "10:00",
    "todos": Array [],
  },
  Object {
    "hour": "11:00",
    "todos": Array [],
  },
  Object {
    "hour": "12:00",
    "todos": Array [],
  },
  Object {
    "hour": "13:00",
    "todos": Array [],
  },
  Object {
    "hour": "14:00",
    "todos": Array [],
  },
  Object {
    "hour": "15:00",
    "todos": Array [],
  },
  Object {
    "hour": "16:00",
    "todos": Array [],
  },
  Object {
    "hour": "17:00",
    "todos": Array [],
  },
  Object {
    "hour": "18:00",
    "todos": Array [],
  },
  Object {
    "hour": "19:00",
    "todos": Array [],
  },
  Object {
    "hour": "20:00",
    "todos": Array [],
  },
  Object {
    "hour": "21:00",
    "todos": Array [],
  },
  Object {
    "hour": "22:00",
    "todos": Array [],
  },
  Object {
    "hour": "23:00",
    "todos": Array [],
  },
  Object {
    "hour": "24:00",
    "todos": Array [],
  },
]
`;
